Add Application centric deployment guide (#453)
theforeman/foreman-documentation: guides/common/modules/con_acd_architecture.adoc

You only need to submit the Ansible playbook to the {ProjectServer}, but not an arbitrary number of connected {SmartProxyServer}s.

[id="{context}_ansible_playbooks"]
== Ansible Playbooks

Ansible playbooks provide the configuration of deployed applications.
They are required along with to an xref:{context}_application_definitions[application definition] to create xref:{context}_application_instances[application instances].

Ansible group variables are mandatory and only supported as `key:value`, i.e. lists and dictionaries are currently not allowed and cannot be edited via management UI.

Ansible playbooks need to contain Ansible group variables for each provided service.

The Ansible inventory is automatically created by the ACD plugin.

In the {ProjectWebUI}, navigate to *Applications > Ansible Playbooks* to manage Ansible playbooks for application centric deployment.

[CAUTION]
====
The Ansible playbook must contain all necessary roles and tasks to execute the playbook.

If your Ansible playbooks are currently using Ansible collections, ensure that these Ansible collections are added manually to every {SmartProxyServer}.

> You can install Ansible collections by hand by running the following command on your {ProjectServer} and {SmartProxyServer}:

Reviewer: ```suggestion
 To install Ansible collections manually, enter the following command on your {ProjectServer} and {SmartProxyServer}:
```
A few things here. 
This "by hand" sounds very strange in English. We generally use the term "manually". 
We always use "enter" for a command and "run" for a script.

Author: Thanks, applied!

[source,bash]
----

> ansible-galaxy collection install namespace.collection

Reviewer: ```suggestion
# ansible-galaxy collection install _namespace.collection_
```

In Foreman docs, we wrap all variables in italics.

Author: Fixed.

[id="{context}_architecture"]
= Architecture

> The ACD plugin is about deploying whole applications instead of single hosts at a time.

Author: Removed.

> ACD uses smart proxies and a new remote execution provider called `acd`.

Author: True! Fixed and checked for other instances of "new".

> Ansible playbooks are downloaded from {ProjectServer} to the {SmartProxyServer} before being executed.

Author: Fixed.

FYI: There are a lot of instances: `git grep "the {SmartProxyServer}"`. Should I resolve this in a future PR?

> Ansible playbooks are downloaded from {ProjectServer} to the {SmartProxyServer} before being executed.

Author: Opened a separate PR: https://github.com/theforeman/foreman-documentation/pull/465

> You only need to submit the Ansible playbook to the {ProjectServer}, but not an arbitrary number of connected {SmartProxyServer}s.

Author: So the idea is that Ansible playbooks are only stored on the ProjectServer and don't have to be sychronizred to any attached SmartProxyServers manually.

This is imporant as in the near future, ACD will also support foreign SmartProxies.

If I understand correctly, other configuration management objects (like Puppet modules etc.) need to by synchronized by hand.

> You only need to submit the Ansible playbook to the {ProjectServer}, but not an arbitrary number of connected {SmartProxyServer}s.

Author: Thanks for your input! Applied.

> This is using a new job provider and new job template.

Author: For users, probably not.

Removed.

[id="{context}_ansible_playbooks"]
== Ansible Playbooks

Ansible playbooks provide the configuration of deployed applications.
They are required next to an xref:{context}_application_definitions[application definition] to create xref:{context}_application_instances[application instances].

Ansible group variables are mandatory and only supported as `key:value`, i.e. lists and dictionaries are currently not allowed and cannot be edited via management UI.

Ansible playbooks need to contain Ansible group variables for each provided service.

The Ansible inventory is automatically created by the ACD plugin.

Navigate to *Applications > Ansible Playbooks* to manage Ansible playbooks for application centric deployment.

[CAUTION]
====

> The Ansible playbook must contain all necessary roles and tasks to execute the playbook.

Author: This is something that has to be proven helpful for new users of ACD.

> Support for https://docs.ansible.com/ansible/latest/user_guide/collections_using.html[Ansible collections] will be added in a future release.

Author: Removed the promise.

> If your Ansible playbooks are currently using Ansible collections, make sure they are present on every attached {SmartProxyServer} by hand.

Author: Applied.

You can install Ansible collections by hand by running the following command on your {ProjectServer} and {SmartProxyServer}:

[source,bash]
----
ansible-galaxy collection install namespace.collection
----

Replace `namespace.collection` with the Ansible collection you are intending to use, e.g. `community.general`.
This uses the Ansible galaxy as configured in the `ansible.cfg` file.
====

[id="{context}_application_definitions"]
== Application Definitions

> Application definitions are blueprints of deployed applications.

Author: Reworded the app def section.

> They are required next to an xref:{context}_ansible_playbooks[Ansible playbook] to create xref:{context}_application_instances[application instances].

Author: Creating an application instance requires an application definition and an Ansible playbook; so "both side by side" are required.

`next to` > `along with`?

> Application definitions are typically handled by administrative {ProjectName} users.

Author: No change.

I would argue this is fine as is; it's an overview of what ACD is and how it's built. User permissions (i.e. necessary user roles) become clear in the `User Roles` section on the same page.

> Application definitions are typically handled by administrative {ProjectName} users.

Reviewer: The main problem with this sentence is how passive it is. For me, documentation should be written in an active voice where at all possible. It speaks to no one.

> Application definitions are typically handled by administrative {ProjectName} users.

Author: Reworded: `Users with administrative access to {ProjectName} handle application definitions and specify the allowed number of hosts for each service.`.

They serve as a _blueprint_ providing easy self service for end users.

> Navigate to *Applications > App Definitions* to manage application definitions for application centric deployment.

Author: Removed `typically` to clarify between user roles.

Application definition cannot be added, edited, or removed by ACD users (users having this Foreman role).

I don't know how users use the plugin in combination with Foreman user roles, but it provides two: ACD Manager and ACD User (only allowed to manage application instances).

Renamed `manage`.

[id="{context}_application_instances"]
== Application Instances

> An application instance is a preconfigured xref:{context}_application_definitions[application definition] and can be rolled out by end users.

Author: Reworded.

This automatically deploys and configures hosts based on the application definition and Ansible playbook.

> Based on one application definition, end users can configure and deploy multiple application instances.

Author: As this is the "architecture" section, we want to make clear that end users only need/can configure an application instance and administrativ users take care of the ansible playbook and application definition.

Therefore, I think it makes sense to use `end users` and `administrative users` which 1:1 correlates to the ACD manager and ACD user roles.

> Navigate to *Applications > App Instances* to manage application instances for application centric deployment.

Author: Reworded to "add, deploy, and remove".

[id="{context}_ansible_variables_and_parameter_hierarchy"]

> == Ansible Variables and Parameter Hierarchy

Author: This section is important to understand how Ansible variables and Foreman parameters are inherited.

I did not find any Foreman documentation on how this works; only [Overriding Ansible Variables](https://www.theforeman.org/plugins/foreman_ansible/2.x/index.html#4.5OverridingAnsibleVariables).

I've added a link to the Ansible docs: [Variable precedence](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable).

Does this make sense?

> == Ansible Variables and Parameter Hierarchy

Author: How should I handle this instead?

There are different levels on how to set Ansible variables and parameters for a service/host.

.Ansible variables
* On the base, there are Ansible group variables for _all_ services as defined in the Ansible playbook.
* Next, there are Ansible group variables defined for _all_ services in the application definition.
Individual Ansible variables can be locked, removed, or added to the application definition.
They overwrite the ones from the Ansible playbook.
* An application definition also contains Ansible variables for _service groups_.
These overwrite Ansible variables set for all services.
* An application instance can overwrite any unlocked Ansible variables from the application definition.

.Parameters
* Managed hosts inherit parameters from multiple entities:
** Global parameters
** Organization parameters
** Location parameters
** Domain parameters
** Subnet parameters
** Operating system parameters
** Host group parameters
** Host parameters

+
For example, domain level parameters overwrite both location and organization based parameters and host parameters overwrite any other defined parameter.

[id="{context}_user_roles"]
== User Roles

> The ACD plugin is used to deploy applications and requires sufficient {ProjectName} permissions.

Author: Applied.

Depending on your environment, there might be two different groups of users in regard to _managing_ an application and _deploying_ an application:

> * An administrative group of users might be responsible for providing the tools to deploy an application, i.e. an Ansible playbook and application definition.

Author: It's more like a "that is" than "for example" in my opinion.
I would argue to leave it as is, as this section specifically addresses user permissions; so `ACD Manager` users have permissions to add, edit, and remove Ansible playbooks and application definitions.

Do you see "for example" because there are more tools necessary to deploy applications?

> * An administrative group of users might be responsible for providing the tools to deploy an application, i.e. an Ansible playbook and application definition.

Author: Fixed.

> * An end user group might be using {ProjectName} to deploy applications, i.e. creating and deploying application instances based on provided application definitions within their organization and location context.

Author: I would not change this because it is part of the architecture section.

I've removed `provided`.

> Many application instances can be created based on the same application definition.

Author: Not removed as this shows that end users can create and deploy multiple applications of the same type.
